File: src/main/java/io/myosotisdev/utopianism/battle/attributes/Health.java
package io.myosotisdev.utopianism.battle.attributes;

import io.myosotisdev.utopianism.battle.ShieldMeta;
import io.myosotisdev.utopianism.util.IntegerTemplate;
import net.minestom.server.entity.LivingEntity;

import java.util.List;

public class Health
{
    private LivingEntity targetingEntity;
    private IntegerTemplate maxValue;
    private IntegerTemplate regen;
    private int actualHealth;
    private List<ShieldMeta> shields;
}

File: src/main/kotlin/io/myosotisdev/utopianism/battle/attributes/Health.kt
package io.myosotisdev.utopianism.battle.attributes

import com.google.common.primitives.Doubles
import io.myosotisdev.utopianism.battle.ShieldMeta
import io.myosotisdev.utopianism.util.IntegerTemplate
import net.minestom.server.entity.LivingEntity

class Health
{
    val entity: LivingEntity? = null
    val maxValue: IntegerTemplate? = null
    val regen: IntegerTemplate? = null
    var actualHealth = 0.0
        set(value)
        {
            field = Doubles.constrainToRange(value, 0.0, maxValue!!.value()
                    .toDouble())
        }
    val shields: List<ShieldMeta>? = null
}